PXC-702: GCache encryption PXC-3063: Record Set cache data at rest encryption

kamil-holubicki opened the pull request:
https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Server part implementation.
1. Needed changes around the keyring component and Galera initialization
2. Master key rotation
3. Master key replication

Integrated existing MTR test suites in the way that they can be executed
with encryption enabled or disabled.

Introduced new Galera encryption MTR tests.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Server part implementation.
1. Needed changes around keyring component and Galera initialization
2. Master key rotation
3. Master key replication

dutow commented on mysql-test/suite/galera/t/galera_gcache_recover_manytrx.test:
```diff
@@ -128,6 +128,9 @@ DELIMITER ;|
 SET SESSION wsrep_sync_wait = 0;
 
+# we need to give it a chance to put something in gcache and in case of encryption for some enc cache pages to be flushed
+--sleep 18
```
Will this also work on jenkins, isn't there a better check?

kamil-holubicki replied:
Yes, it works on Jenkins.

This magic --sleep 18 is here a few lines below in the original test.
The flow of the original test is:
1. Start transactions on node_1
2. Wait on node_2 just for CREATE TABLE to be replicated (and probably some inserts, but do not care how much of them)
3. kill node_2
4. wait for 18 secs. node_1 still processes inserts
5. start node_2 - it should do gcache recovery and join with SST.

There is a problem with points 2 and 3 when we use gcache encryption. As encryption uses in-memory cache which is being flushed only when there are no more cache pages available, if we kill immediately in point 3, there is nothing in GCache file. So the recovery will not take place.
We need to wait for some time for GCache encryption cache to be flushed internally (and this will be done if we receive more transactions from node_1). Then, after node_2 restart it will see that there is "something" in gcache file and the recovery will take place.
Why 18 secs? well, because the original test used 18 as well :).  There is no way to say that the encryption cache has been flushed (the only way is to examine gcache file itself, but we still would need to wait for some arbitrary time, check and then wait again if check fails...) but I didn't want to change this test too much.

dutow commented on mysql-test/suite/galera_encryption/t/gcache_mk_rotation.test:
```diff
@@ -0,0 +1,90 @@
+--let $assert_count = 1
+--source include/assert_grep.inc
+
+ALTER INSTANCE ROTATE GCACHE MASTER KEY;
```
What if alter instance ... is executed without wsrep_on, should that work, and rotate the key only on one server?

Is it possible that one node uses gcache encryption, the other doesn't?

kamil-holubicki replied:
This is not wsrep_on -aware. It works the same way as ALTER INSTANCE ROTATE INNODB MASTER KEY works.

If GCache encryption is not enabled, it does not rotate the key (because enabling gcache encryption needs server restart and gcache reset, so no point in rotating the key if it will not be used anyway), but request for rotation is replicated, because another servers can have encryption enabled.

I will add test scenario for this.

kamil-holubicki replied:
MTR test added

kamil-holubicki replied:
Please forget my above explanation. I messed things up :D. 
WSREP_ON = OFF prevents replication. rotation is done on local node and is not replicated. The same as for InnoDB MK. Added MTR for this.

dutow commented on sql/sql_plugin.cc:
```diff
@@ -1581,25 +1584,45 @@ bool update_persisted_plugin_sysvars(const char *name) {
 
   /* First we register the builtin mandatory and optional plugins */
-  for (struct st_mysql_plugin **builtins = mysql_mandatory_plugins;
+  for (struct st_mysql_plugin **builtins = plugins /*mysql_mandatory_plugins*/;
```
missing ifdef wsrep - and with this change, the above comment is also incorrect.

kamil-holubicki replied:
fixed

dutow commented on sql/sys_vars.cc:
```diff
@@ -8708,6 +8708,20 @@ static Sys_var_bool Sys_pxc_encrypt_cluster_traffic(
+static const char *wsrep_encrypt_modes[] = {"OFF", "ON", "NONE",
+                                        NullS};
+static Sys_var_enum Sys_wsrep_gcache_encrypt(
+    "wsrep_gcache_encrypt", "Encrypt GCache. This variable is solely for testing purposes. "
```
Do we plan to keep this variable when we release it? Shouldn't it be behind some testing/internal ifdef?

kamil-holubicki replied:
I think it is OK to have it here. I would even not document it. It will not do any harm and in worst-case scenario will act as the override of wsrep_provider_params' gcache.encrypt parameter.
Having it in release allows us to run standard MTR suites

dutow commented on sql/wsrep_master_key_manager.cc:
```diff
@@ -0,0 +1,153 @@
+  /* Key absent */
+  if (reader_object == nullptr) return std::string();
+
+  auto cleanup_guard = create_scope_guard([&] {
```
Shouldn't this guard be somewhere earlier, like around the definition of reader_object?

kamil-holubicki replied:
This guard is to deinit the `keyring_reader_service` which was properly inited. 
We know that it was properly inited if 
1. retval != true (1st `if`, if retval == true, it is not said that keyring_reader_service is untouched.
2. `keyring_reader_service` is not null (2nd if)

So I would say we can move it before 2nd if, but for sure not before the 1st one.
I would also say, that current implementation is fine :). The implementation is the same as keyring_operations_helper.cc::read_secret()

Do you agree?

dutow replied:
Yes, it seems to be fine with the current components/plugins we have. 

My thought was: what if a keyring sets the reader object, but returns  an error, shouldn't we deinit it in that case? Of course we don't have anything like that, so this isn't an actual issue for PXC.

dutow reviewed (commented):
Is this pull request a duplicate of #1660?

kamil-holubicki commented:
>Is this pull request a duplicate of https://github.com/percona/percona-xtradb-cluster/pull/1660?
Not exactly. This is reworked version (based on that one), but a lot of squashing, reordering... I decided to go with new one as it was easier/quicker. 
Closing the other one.

dutow commented on sql/wsrep_mysqld.cc:
```diff
@@ -3122,3 +3180,29 @@ bool wsrep_consistency_check(THD *thd) {
+}
+
+std::string wsrep_get_master_key(const std::string &keyId) {
+  std::string mk = masterKeyManager->GetKey(keyId);
```
Is this variable needed, wouldn't a simple return work?

kamil-holubicki replied:
done

dutow commented on sql/wsrep_master_key_manager.cc:
```diff
@@ -0,0 +1,153 @@
+}
+
+bool MasterKeyManager::GenerateKey(const std::string &keyId) {
+  if (keyring_generator_service_->generate(keyId.c_str(), nullptr, key_type,
```
Could be return ... instead of if?

kamil-holubicki replied:
done

dutow commented on sql/wsrep_master_key_manager.cc:
```diff
@@ -0,0 +1,153 @@
+  /* Allocate requried memory for key and secret_type */
+  std::vector<unsigned char> secret_v;
+  std::vector<char> secret_type_v;
+  secret_v.reserve(secret_length);
```
Why reserve and not resize? That would also take care of the memset.

kamil-holubicki replied:
done

dutow commented on sql/wsrep_mysqld.cc:
```diff
@@ -1066,6 +1073,50 @@ void wsrep_deinit_server() {
+    override_galera_option("allocator.disk_pages_encryption", val, options);
+  }
+
+  // todo: strncpy, buffer size, etc
```
leftover todo

kamil-holubicki replied:
done

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Integrated existing MTR test suites in the way that they can be executed
with encryption enabled or disabled. Used combinations file for this and
two auxiliary read-only variables:
wsrep_gcache_encrypt
wsrep_disk_pages_encrypt

Using of these variables is necessary, because wserp_provider_params
cannot be extended via combinations file.

wsrep_deinit() call added to unireg_abort() (MK infrastructure
deinitialization)

SST channel is encrypted during MTR tests, as it is necessary when
keyring plugin is loaded.

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Galera encryption MTR tests

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Addressed review comments

kamil-holubicki committed:
PXC-702: GCache encryption
PXC-3063: Record Set cache data at rest encryption

https://jira.percona.com/browse/PXC-702
https://jira.percona.com/browse/PXC-3063

Galera and wsrep-lib submodules pointers updated.
